osquery/tables/system/windows/services.cpp

```diff
@@ -0,0 +1,164 @@
+
+  if (schService == nullptr) {
+    TLOG << "OpenService failed (" << GetLastError() << ")";
+    return FALSE;
+  }
+
+  QueryServiceConfig(schService, nullptr, 0, &cbBufSize);
+  auto lpsc = (LPQUERY_SERVICE_CONFIG)malloc(cbBufSize);
+  if (!QueryServiceConfig(schService, lpsc, cbBufSize, &cbBufSize)) {
+    TLOG << "QueryServiceConfig failed (" << GetLastError() << ")";
+  }
+
+  QueryServiceConfig2(
+      schService, SERVICE_CONFIG_DESCRIPTION, nullptr, 0, &cbBufSize);
+  auto lpsd = (LPSERVICE_DESCRIPTION)malloc(cbBufSize);
+  if (!QueryServiceConfig2(schService,
+                           SERVICE_CONFIG_DESCRIPTION,
+                           (LPBYTE)lpsd,
+                           cbBufSize,
+                           &cbBufSize)) {
+    TLOG << "QueryServiceConfig2 failed (" << GetLastError() << ")";
+  }
+
+  r["name"] = SQL_TEXT(svc.lpServiceName);
+  r["display_name"] = SQL_TEXT(svc.lpDisplayName);
+  r["status"] = SQL_TEXT(kSvcStatus[svc.ServiceStatusProcess.dwCurrentState]);
+  r["pid"] = INTEGER(svc.ServiceStatusProcess.dwProcessId);
+  r["win32_exit_code"] = INTEGER(svc.ServiceStatusProcess.dwWin32ExitCode);
+  r["service_exit_code"] =
+      INTEGER(svc.ServiceStatusProcess.dwServiceSpecificExitCode);
+  r["start_type"] = SQL_TEXT(kSvcStartType[lpsc->dwStartType]);
+  r["path"] = SQL_TEXT(lpsc->lpBinaryPathName);
+  r["user_account"] = SQL_TEXT(lpsc->lpServiceStartName);
+
+  if (lpsd->lpDescription != nullptr) {
+    r["description"] = SQL_TEXT(lpsd->lpDescription);
+  }
+
+  if (kServiceType.count(lpsc->dwServiceType) > 0) {
+    r["service_type"] = SQL_TEXT(kServiceType.at(lpsc->dwServiceType));
+  } else {
+    r["service_type"] = SQL_TEXT("UNKNOWN");
+  }
+
+  QueryData regResults;
+  queryKey("HKEY_LOCAL_MACHINE",
+           "SYSTEM\\CurrentControlSet\\Services\\" + r["name"] + "\\Parameters",
+           regResults);
+  for (const auto& aKey : regResults) {
+    if (aKey.at("name") == "ServiceDll") {
+      r["module_path"] = SQL_TEXT(aKey.at("data"));
+    }
+  }
+
+  free(lpsc);
+  free(lpsd);
+  CloseServiceHandle(schService);
+  return TRUE;
+}
+
+QueryData genServices(QueryContext& context) {
+  void* buf = nullptr;
+  DWORD bytesNeeded = 0;
+  DWORD serviceCount = 0;
+  QueryData results;
+
+  auto schScManager = OpenSCManager(nullptr, nullptr, GENERIC_READ);
+  if (schScManager == nullptr) {
+    TLOG << "EnumServiceStatusEx failed (" << GetLastError() << ")";
+    return {};
+  }
+
+  (void)EnumServicesStatusEx(schScManager,
+                             SC_ENUM_PROCESS_INFO,
+                             SERVICE_WIN32,
+                             SERVICE_STATE_ALL,
+                             nullptr,
+                             0,
+                             &bytesNeeded,
+                             &serviceCount,
+                             nullptr,
+                             nullptr);
+
+  buf = malloc(bytesNeeded);
```

Comment:
If you nix the declaration of `buf` above you can `auto buf = (LPBYTE)malloc(bytesNeeded);` here-- then not have to cast withing the arguments provided to `EnumServicesStatusEx` below.